pageserver/src/tenant/storage_layer/delta_layer.rs

```diff
@@ -549,25 +549,13 @@ impl DeltaLayer {
             &self.layer_name(),
         )
     }
-
-    /// Obtains all keys and value references stored in the layer
+    /// Loads all keys stored in the layer. Returns key, lsn, value size and value reference.
     ///
     /// The value can be obtained via the [`ValueRef::load`] function.
-    pub async fn load_val_refs(
+    pub async fn load_keys(
         &self,
         ctx: &RequestContext,
-    ) -> Result<Vec<(Key, Lsn, ValueRef<Arc<DeltaLayerInner>>)>> {
-        let inner = self
-            .load(LayerAccessKind::Iter, ctx)
-            .await
-            .context("load delta layer")?;
-        DeltaLayerInner::load_val_refs(inner)
-            .await
-            .context("Layer index is corrupted")
-    }
-
-    /// Loads all keys stored in the layer. Returns key, lsn and value size.
-    pub async fn load_keys(&self, ctx: &RequestContext) -> Result<Vec<(Key, Lsn, u64)>> {
+    ) -> Result<Vec<(Key, Lsn, u64, ValueRef<Arc<DeltaLayerInner>>)>> {
```

Comment:
The return type is complicated enough that a new struct might be in order.

Reply:
Not new with this PR, but do we bump the reference count on the `Arc<DeltaLayerInner>` for every ValueRef? If so, we could probably save a lot of CPU cycles by not doing that.

Reply:
I introduced this scheme in #4836. The reference count is just an atomic counter, changing it should be pretty cheap, but of course has some synchronization overhead if edited between cpus.

There are a bunch of alternatives, but imo neither are as nice as using the `Arc` pointer:

* using unsafe plus a raw pointer to the `DeltaLayerInner`
* changing `ValueRef::load` to take a `DeltaLayerInner` by reference, aka moving the handling outside. Then one introduces potential wrong usage when the wrong `ValueRef` is used with the wrong `DeltaLayer`.
* I might be misremembering, but I think using references doesn't work because of async and await points (holding references across await points means the pointed thing must be Sync otherwise the future is not Send). Not 100% sure though, maybe worth a try.

Reply:
> I think using references doesn't work

I made some some experiments, and it seems that one can easily avoid cloning arcs here. I was mistaken! This is great news, we can just use references. I will add a commit.

Reply:
I've added commits for both using references and switching to a `DeltaEntry` struct.

Reply:
This was a really annoying change to rebase over.